### PR TITLE
replace many unnecessary gsub calls with sub

### PR DESCRIPTION
**Pull request by phluid61**
It doesn't make sense to globally replace an anchored expression.

### CODE REVIEW

**Comment by phluid61** [comment]
I mean, sure, Hound Bot, I saw that while I was editing... but one change at a time, eh?

File: lib/addressable/uri.rb [hunk 1]
```diff
@@ -1900,7 +1900,7 @@ def join(uri)
               #
               # Removes the right-most path segment from the base path.
               if base_path =~ /\//
-                base_path.gsub!(/\/[^\/]+$/, SLASH)
+                base_path.sub!(/\/[^\/]+$/, SLASH)
```
[the remark this answers is not in the document]
Comment (sporkmonger) [reply lib/addressable/uri.rb]: Agree w/ hound here.

File: lib/addressable/uri.rb [hunk 2]
```diff
@@ -1544,7 +1544,7 @@ def path=(new_path)
     # @return [String] The path's basename.
     def basename
       # Path cannot be nil
-      return File.basename(self.path).gsub(/;[^\/]*$/, EMPTY_STR)
+      return File.basename(self.path).sub(/;[^\/]*$/, EMPTY_STR)
```
[the remark this answers is not in the document]
Comment (sporkmonger) [reply lib/addressable/uri.rb]: Agree with hound on using `%r`. Ignore the other two for now.

File: lib/addressable/uri.rb [hunk 3]
```diff
@@ -263,19 +263,19 @@ def self.convert_path(path)
...
         # Adjust windows-style uris
-        uri.path.gsub!(/^\/?([a-zA-Z])[\|:][\\\/]/) do
+        uri.path.sub!(/^\/?([a-zA-Z])[\|:][\\\/]/) do
           "/#{$1.downcase}:/"
         end
         uri.path.gsub!(/\\/, SLASH)
         if File.exist?(uri.path) &&
             File.stat(uri.path).directory?
-          uri.path.gsub!(/\/$/, EMPTY_STR)
+          uri.path.sub!(/\/$/, EMPTY_STR)
```
[the remark this answers is not in the document]
Comment (sporkmonger) [reply lib/addressable/uri.rb]: Agree.

File: lib/addressable/uri.rb [hunk 4]
```diff
@@ -263,19 +263,19 @@ def self.convert_path(path)
...
 
-      path.gsub!(/^file:\/?\/?/, EMPTY_STR) if path =~ /^file:\/?\/?/
+      path.sub!(/^file:\/?\/?/, EMPTY_STR) if path =~ /^file:\/?\/?/
       path = SLASH + path if path =~ /^([a-zA-Z])[\|:]/
       uri = self.parse(path)
 
       if uri.scheme == nil
         # Adjust windows-style uris
-        uri.path.gsub!(/^\/?([a-zA-Z])[\|:][\\\/]/) do
+        uri.path.sub!(/^\/?([a-zA-Z])[\|:][\\\/]/) do
```
[the remark this answers is not in the document]
Comment (sporkmonger) [reply lib/addressable/uri.rb]: Agree.

File: lib/addressable/uri.rb [hunk 5]
```diff
@@ -263,19 +263,19 @@ def self.convert_path(path)
       # Otherwise, convert to a String
       path = path.to_str.strip
 
-      path.gsub!(/^file:\/?\/?/, EMPTY_STR) if path =~ /^file:\/?\/?/
+      path.sub!(/^file:\/?\/?/, EMPTY_STR) if path =~ /^file:\/?\/?/
```
[the remark this answers is not in the document]
Comment (sporkmonger) [reply lib/addressable/uri.rb]: Agree.

File: lib/addressable/uri.rb [hunk 6]
```diff
@@ -182,18 +182,18 @@ def self.heuristic_parse(uri, hints={})
...
+      when /^http:\//
+        uri.sub!(/^http:\/+/, "http://")
+      when /^https:\//
+        uri.sub!(/^https:\/+/, "https://")
+      when /^feed:\/+http:\//
+        uri.sub!(/^feed:\/+http:\/+/, "feed:http://")
+      when /^feed:\//
+        uri.sub!(/^feed:\/+/, "feed://")
+      when /^file:\//
+        uri.sub!(/^file:\/+/, "file:///")
```
[the remark this answers is not in the document]
Comment (sporkmonger) [reply lib/addressable/uri.rb]: Agree.

File: lib/addressable/uri.rb [hunk 7]
```diff
@@ -182,18 +182,18 @@ def self.heuristic_parse(uri, hints={})
...
-        uri.gsub!(/^file:\/+/, "file:///")
+      when /^http:\//
+        uri.sub!(/^http:\/+/, "http://")
+      when /^https:\//
+        uri.sub!(/^https:\/+/, "https://")
+      when /^feed:\/+http:\//
+        uri.sub!(/^feed:\/+http:\/+/, "feed:http://")
+      when /^feed:\//
+        uri.sub!(/^feed:\/+/, "feed://")
+      when /^file:\//
```
[the remark this answers is not in the document]
Comment (sporkmonger) [reply lib/addressable/uri.rb]: Agree.

File: lib/addressable/uri.rb [hunk 8]
```diff
@@ -182,18 +182,18 @@ def self.heuristic_parse(uri, hints={})
...
-      when /^file:\/+/
-        uri.gsub!(/^file:\/+/, "file:///")
+      when /^http:\//
+        uri.sub!(/^http:\/+/, "http://")
+      when /^https:\//
+        uri.sub!(/^https:\/+/, "https://")
+      when /^feed:\/+http:\//
+        uri.sub!(/^feed:\/+http:\/+/, "feed:http://")
+      when /^feed:\//
+        uri.sub!(/^feed:\/+/, "feed://")
```
[the remark this answers is not in the document]
Comment (sporkmonger) [reply lib/addressable/uri.rb]: Agree.

File: lib/addressable/uri.rb [hunk 9]
```diff
@@ -182,18 +182,18 @@ def self.heuristic_parse(uri, hints={})
...
-        uri.gsub!(/^feed:\/+/, "feed://")
-      when /^file:\/+/
-        uri.gsub!(/^file:\/+/, "file:///")
+      when /^http:\//
+        uri.sub!(/^http:\/+/, "http://")
+      when /^https:\//
+        uri.sub!(/^https:\/+/, "https://")
+      when /^feed:\/+http:\//
+        uri.sub!(/^feed:\/+http:\/+/, "feed:http://")
+      when /^feed:\//
```
[the remark this answers is not in the document]
Comment (sporkmonger) [reply lib/addressable/uri.rb]: Agree.

File: lib/addressable/uri.rb [hunk 10]
```diff
@@ -182,18 +182,18 @@ def self.heuristic_parse(uri, hints={})
...
-      when /^feed:\/+/
-        uri.gsub!(/^feed:\/+/, "feed://")
-      when /^file:\/+/
-        uri.gsub!(/^file:\/+/, "file:///")
+      when /^http:\//
+        uri.sub!(/^http:\/+/, "http://")
+      when /^https:\//
+        uri.sub!(/^https:\/+/, "https://")
+      when /^feed:\/+http:\//
+        uri.sub!(/^feed:\/+http:\/+/, "feed:http://")
```
[the remark this answers is not in the document]
Comment (sporkmonger) [reply lib/addressable/uri.rb]: Agree.

File: lib/addressable/uri.rb [hunk 11]
```diff
@@ -182,18 +182,18 @@ def self.heuristic_parse(uri, hints={})
...
-        uri.gsub!(/^feed:\/+http:\/+/, "feed:http://")
-      when /^feed:\/+/
-        uri.gsub!(/^feed:\/+/, "feed://")
-      when /^file:\/+/
-        uri.gsub!(/^file:\/+/, "file:///")
+      when /^http:\//
+        uri.sub!(/^http:\/+/, "http://")
+      when /^https:\//
+        uri.sub!(/^https:\/+/, "https://")
+      when /^feed:\/+http:\//
```
[the remark this answers is not in the document]
Comment (sporkmonger) [reply lib/addressable/uri.rb]: Agree.

File: lib/addressable/uri.rb [hunk 12]
```diff
@@ -182,18 +182,18 @@ def self.heuristic_parse(uri, hints={})
...
-        uri.gsub!(/^https:\/+/, "https://")
-      when /^feed:\/+http:\/+/
-        uri.gsub!(/^feed:\/+http:\/+/, "feed:http://")
-      when /^feed:\/+/
-        uri.gsub!(/^feed:\/+/, "feed://")
-      when /^file:\/+/
-        uri.gsub!(/^file:\/+/, "file:///")
+      when /^http:\//
+        uri.sub!(/^http:\/+/, "http://")
+      when /^https:\//
```
[the remark this answers is not in the document]
Comment (sporkmonger) [reply lib/addressable/uri.rb]: Agree.

File: lib/addressable/uri.rb [hunk 13]
```diff
@@ -182,18 +182,18 @@ def self.heuristic_parse(uri, hints={})
...
-      when /^https:\/+/
-        uri.gsub!(/^https:\/+/, "https://")
-      when /^feed:\/+http:\/+/
-        uri.gsub!(/^feed:\/+http:\/+/, "feed:http://")
-      when /^feed:\/+/
-        uri.gsub!(/^feed:\/+/, "feed://")
-      when /^file:\/+/
-        uri.gsub!(/^file:\/+/, "file:///")
+      when /^http:\//
+        uri.sub!(/^http:\/+/, "http://")
```
[the remark this answers is not in the document]
Comment (sporkmonger) [reply lib/addressable/uri.rb]: Agree.

**Comment by sporkmonger** [comment]
I'm fine w/ skipping these style changes to keep the PR more focused though.

**Pull request merged**